### PR TITLE
Allow alternate event/hook prototype declarations

### PR DESCRIPTION
**Pull request by jsiwek**
These changes allow multiple, alternate prototype declarations for events/hooks, similar to what was outlined from https://github.com/zeek/zeek/issues/418#issuecomment-600497140 (though we can still keep that issue open to pertain to adding the feature of true *function* overloads... that's less useful and lower priority).   Primary benefit of this feature is it allows us a way to extend or change existing events/hooks in a way that won't break user handlers (see commit message/docs for examples).

Need to also merge same branch name in `zeek-docs` and `bifcl` along with this PR:
* https://github.com/zeek/zeek-docs/pull/25
* https://github.com/zeek/bifcl/pull/4

### CODE REVIEW

File: src/Var.cc [hunk 1]
```diff
@@ -351,16 +445,24 @@ void begin_func(ID* id, const char* module_name, function_flavor flavor,
...
 	if ( id->Type() )
 		{
-		if ( ! same_type(id->Type(), t.get()) )
-			id->Type()->Error("incompatible types", t.get());
+		prototype = func_type_check(id->Type()->AsFuncType(), t.get());
 
-		else
+		if ( prototype )
+			{
 			// If a previous declaration of the function had &default params,
```
Comment (rsmmr) [inline src/Var.cc:456]: Does this derive default arguments only from the canonical version? (Which would seem like the right behaviour)

This is something worth both documenting and testing I think, mind adding that?

**Comment by rsmmr** [comment]
Looks good, just one comment added

File: src/Var.cc (hunk 1 above)
Comment (jsiwek) [reply src/Var.cc]: Yeah, `&default` args only come through via the canonical event/hook prototype.  Added docs/tests and found a few other places that warranted additional checks.  Let me know if you look and think of anything else or want me to go ahead and merge.

**Comment by rsmmr** [comment]
took a quick look, lgtm, fine with me to go ahead and merge

**Pull request merged**